### PR TITLE
Neuron part two

### PR DESCRIPTION
**Pull request by jimmykodes**

### CODE REVIEW

File: _posts/2019-07-16-neuron.md [hunk 1]
```diff
@@ -155,4 +155,5 @@ After running this a few times, the distribution of -1 and 1 was about equal. Th
 mean this prediction is pretty much coin toss, at least until we train our neuron.
 
 ### To be continued...
-Next week we'll return with a tale of how to train your dra... neuron. How to train your neuron.
+In the [next post]({% post_url 2019-07-31-neuron2 %}), we'll return with a tale of how to train your dra... neuron. 
+How to train your neuron.
```
Comment (ckcollab) [inline _posts/2019-07-16-neuron.md:159]: :D

File: _posts/2019-07-31-neuron2.md [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+Running this code a dozen times should yield a pretty random split of 1s and 0s. The goal of this post is to reduce
+that randomness it steer it closer and closer to a correct prediction.
+
+# Forward the Foundation
+
+### Baseline
+To start, I created a bunch of random data for my perceptron to work with. Without doing any training, I ran it through
+several attempts at classifying the data. Here are a few of its outcomes:
+<div>
+    <img src="/assets/images/articles/initial_performance.png">
```
Comment (ckcollab) [inline _posts/2019-07-31-neuron2.md:54]: I've been following this pattern for images:

```
<div style="text-align: center;">
    <img src="/assets/images/articles/sprint-process.png" class="img-bordered">
</div>
```

Gives them a lil border, centers if not full width, etc. -- may not be super useful in this case, kind of annoying I have to look this snippet up each time as well..

File: _posts/2019-07-31-neuron2.md [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+closer to accurate predictions.
+
+### Step six: training
+Our perceptron is going to go through supervised learning to train. This means we train it on a subset of data for which
+we know the expected outcome. Are the inputs above, class A or class B? Our neuron doesn't know, so we need
+to tell it. It can then look at what it predicted, and what it should have predicted, and make some changes to itself to
+compensate. The only adjustments this neuron can make are to its weights, and we want it to be making adjustments based
+on its error. To calculate this error, we subtract the predicted value from the expected value, which gives us
+these possible error values:
+<table class="ui celled table">
```
Comment (ckcollab) [inline _posts/2019-07-31-neuron2.md]: I would do it like this, unless we need special styling on the table maybe?

```
| Desired | Predicted | Error |
|--------:|-----------|-------|
|       1 | 1         | 0     |
|       1 | -1        | 2     |
|      -1 | -1        | 0     |
|      -1 | 1         | -2    |
```

File: _posts/2019-07-31-neuron2.md [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+        self.weights = weights if weights is not None else self._random_weights()
+        self.training_rate = training_rate
+```
+This makes it easy to play around with training rates without having to change my base neuron code each time.
+We have the ability to adjust our weights, now it's time to flex those neuron muscles... brains have muscles, right? 
+No... moving on...
+
+#### Put in reps
+We are going to make a function called, predictably, `train`. This function could work a couple of ways, it could take
+in a single set of inputs, the single expected output, make a prediction, and then adjust its weights, only doing this
```
Comment (ckcollab) [inline _posts/2019-07-31-neuron2.md]: Could maybe use bullets here to split up this large paragraph?

--------

We are going to make a function called, predictably, train. This function could work a couple of ways:

 * it could take in a single set of inputs, the single expected output, make a prediction, and then adjust its weights, only doing this once per call
 * or, we could pass it our whole set of training data and let it train on all of it. 

For this neuron, I’m choosing to go with a step training process, feeding it the training data one piece at a time. This will make my implementation of my neuron slightly more cumbersome, since I’ll have to handle the data looping there, but it will allow me to do some interesting things when we get to testing.

File: _posts/2019-07-31-neuron2.md [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+    # ...
+    def train(self, inputs, label):
+        output = self.predict(inputs)
+        error = label - output
+        self._adjust_weights(inputs, error)
+```
+Now that we can train, lets do it, and see what we get!
+
+<div>
+    <img src="/assets/images/articles/trained_performance.png">
```
Comment (ckcollab) [inline _posts/2019-07-31-neuron2.md:137]: same img comment as above

File: _posts/2019-07-31-neuron2.md [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+<div>
+    <img src="/assets/images/articles/final_run.png">
+</div>
+The plot on the left is the error rate over time, and the plot on the right is the final test result, after all the
+training has been done. Feel free to play with all this yourself. All of the code, plus some supplementary bits can
+be found [here](https://github.com/jimmykodes/neuron_blog_post).
+
+## The whole enchilada. Thanks, now I want dinner!
+Here it is. The whole kit and caboodle. Lock, stock, and barrel. Hook, line, and sinker. You get the idea...
+```python
```
Comment (ckcollab) [inline _posts/2019-07-31-neuron2.md:207]: cool :)
